File: langs/python/src/vmx/notifications/null_notification_hub.py
"""NullNotificationHub — null-object variant. See ADR-0017 and ADR-0013."""

from __future__ import annotations

import asyncio

import reactivex as rx
from reactivex import Observable

from vmx.notifications.notification import Notification
from vmx.notifications.notification_reaction import NotificationReaction


class NullNotificationHub:
    """Null-object hub: post resolves Approve immediately; resolve is a no-op."""

    def __init__(self) -> None:
        self._pending: Observable[list[Notification]] = rx.from_iterable([[]])

    @property
    def pending(self) -> Observable[list[Notification]]:
        return self._pending

    def post(self, notification: Notification) -> asyncio.Future[NotificationReaction]:
        loop = asyncio.get_running_loop()
        future: asyncio.Future[NotificationReaction] = loop.create_future()
        future.set_result(NotificationReaction.APPROVE)
        return future

    def resolve(self, notification: Notification, reaction: NotificationReaction) -> None:
        return None


NULL_NOTIFICATION_HUB: NullNotificationHub = NullNotificationHub()
"""Shared singleton."""

File: langs/swift/Sources/VMx/Notifications/NullNotificationHub.swift
//
// NullNotificationHub.swift — null-object variant per ADR-0017.
//
// See spec/16-notifications.md §3 and ADR-0017.
//
import Combine

/// Null-object `NotificationHubProtocol`: `post` returns `.approve` immediately
/// without suspension, `resolve` is a no-op, and `pending` is always empty.
///
/// Conforms to ADR-0017 (null-object pattern). The singleton `INSTANCE` is
/// provided for convenience; `public init()` matches `NullLocalizer` precedent
/// so it is injectable wherever a `NotificationHubProtocol` is required.
public final class NullNotificationHub: NotificationHubProtocol {
    /// Shared singleton instance. The hub is stateless.
    public static let INSTANCE = NullNotificationHub()

    public init() {}

    /// Always-empty pending snapshot; completes immediately on subscribe.
    public let pending: AnyPublisher<[Notification], Never> =
        Just([Notification]()).eraseToAnyPublisher()

    /// Returns `.approve` immediately without suspension.
    public func post(_ n: Notification) async -> NotificationReaction {
        return .approve
    }

    /// No-op: there are no pending notifications to resolve.
    public func resolve(_ n: Notification, _ reaction: NotificationReaction) {
        // intentional no-op
    }
}
